packages/nx/src/daemon/server/sync-generators.ts

```diff
@@ -0,0 +1,245 @@
+    if (waitPeriod < 4000) {
+      waitPeriod = waitPeriod * 2;
+    }
+
+    if (scheduledGenerators.size === 0) {
+      // no generators to run
+      return;
+    }
+
+    const { projects } =
+      readProjectsConfigurationFromProjectGraph(projectGraph);
+
+    for (const generator of scheduledGenerators) {
+      runGenerator(generator, projects);
+    }
+
+    await Promise.all(syncGeneratorsCacheResultPromises.values());
+  }, waitPeriod);
+}
+
+export async function getCachedRegisteredSyncGenerators(): Promise<string[]> {
+  log('get registered sync generators');
+  if (!registeredSyncGenerators) {
+    log('no registered sync generators, collecting them');
+    const { projectGraph } = await getCachedSerializedProjectGraphPromise();
+    collectAllRegisteredSyncGenerators(projectGraph);
+  } else {
+    log('registered sync generators already collected, returning them');
+  }
+
+  return [...registeredSyncGenerators];
+}
+
+function collectAllRegisteredSyncGenerators(projectGraph: ProjectGraph): void {
+  const projectGraphHash = hashProjectGraph(projectGraph);
+  if (storedProjectGraphHash !== projectGraphHash) {
+    storedProjectGraphHash = projectGraphHash;
+    registeredTaskSyncGenerators =
+      collectRegisteredTaskSyncGenerators(projectGraph);
+  } else {
+    log('project graph hash is the same, not collecting task sync generators');
+  }
+
+  const nxJson = readNxJson();
+  const nxJsonHash = hashArray(nxJson.sync?.globalGenerators?.sort() ?? []);
+  if (storedNxJsonHash !== nxJsonHash) {
+    storedNxJsonHash = nxJsonHash;
+    registeredGlobalSyncGenerators =
+      collectRegisteredGlobalSyncGenerators(nxJson);
+  } else {
+    log('nx.json hash is the same, not collecting global sync generators');
+  }
+
+  const generators = new Set([
+    ...registeredTaskSyncGenerators,
+    ...registeredGlobalSyncGenerators,
+  ]);
+
+  if (!registeredSyncGenerators) {
+    registeredSyncGenerators = generators;
+    return;
+  }
+
+  for (const generator of registeredSyncGenerators) {
+    if (!generators.has(generator)) {
+      registeredSyncGenerators.delete(generator);
+      syncGeneratorsCacheResultPromises.delete(generator);
+    }
+  }
+
+  for (const generator of generators) {
+    if (!registeredSyncGenerators.has(generator)) {
+      registeredSyncGenerators.add(generator);
+    }
+  }
+}
+
+function runGenerator(
+  generator: string,
+  projects: Record<string, ProjectConfiguration>
+): void {
+  log('running scheduled generator', generator);
+  // remove it from the scheduled set
+  scheduledGenerators.delete(generator);
+  const tree = new FsTree(
+    workspaceRoot,
+    false,
+    `running sync generator ${generator}`
+  );
+
+  // run the generator and cache the result
+  syncGeneratorsCacheResultPromises.set(
+    generator,
+    runSyncGenerator(tree, generator, projects).then((result) => {
+      log(generator, 'changes:', result.changes.map((c) => c.path).join(', '));
+      return result;
+    })
```

Comment:
Does this handle when the sync generator is not async?

Reply:
It does. The `runSyncGenerator` is an async function that'll always return a `Promise`. I've also tested it with a mocked sync version of the `typescript-sync` generator.